### DISCUSSION TITLE
Allow Webhook to watch a specific storage bucket rather than an entire table

### DISCUSSION THREAD

**Discussion by danvoyce** [supabase/supabase#22879] [Feature Requests]
I have a Webhook that is currently being triggered when an insert happens on `storage objects`. The edge function that is being called is actually modifying the image and inserting it into a different storage bucket, and thus triggering the edge function again. I'm having to do the following check `record.bucket_id !== SOURCE_BUCKET_ID` to prevent an infinite loop. I could also setup a separate project to house the source objects, however it would be more elegant / simpler if I could select a bucket rather than the whole table. 

Many thanks and amazing work as always!

<img width="771" alt="Screenshot 2024-04-18 at 17 11 46" src="https://github.com/supabase/supabase/assets/1253802/dc42139c-8fcf-4149-a2c3-f1f29c470fdd">

**Comment by GaryAustin1** [comment]
You could just use your own trigger function and the WHEN clause to filter and then call a postgres function and use the pg_net or http extension to call your edge function.

**Reply by danvoyce** [reply]
I'm struggling to get the query right. This is what I've got so far:

```sql
CREATE OR REPLACE FUNCTION image_processing() RETURNS TRIGGER AS $$
BEGIN
  PERFORM supabase_functions.http_request(
    url => concat(current_setting('SUPABASE_URL', true), '/functions/v1/handle_trip_photo_upload'),
    method => 'POST',
    headers => '{"Content-Type": "application/json"}',
    body => '{}',
    timeout => '5000'
  );
  RETURN NEW;
END;
$$
 LANGUAGE plpgsql;

CREATE OR REPLACE TRIGGER trigger_image_processing
AFTER INSERT ON storage.objects
FOR EACH ROW WHEN (NEW.bucket_id = 'trip-photos-originals')
EXECUTE FUNCTION trigger_image_processing();
```

But now when I upload an image I get this error:

`insert into "objects" ("bucket_id", "metadata", "name", "owner", "owner_id", "version") values ($1, DEFAULT, $2, $3, $4, $5) - invalid input syntax for type json`

Any idea?

**Reply by GaryAustin1** [reply]
I've not use supabase_functions.http_request directly myself.
Maybe you need to cast headers and body with '{}'::jsonb?

Also have you tested current_setting('SUPABASE_URL',true)?  I've not seen that and selecting it on my instance returns null.

**Reply by tomsoderlund** [reply]
@danvoyce Did you solve your trigger query?

**Reply by tomsoderlund** [reply]
…and shouldn’t the last line be:

    EXECUTE FUNCTION image_processing();

to avoid self-referencing?

**Comment by tomsoderlund** [comment]
For new people coming here, I believe **Supabase Database Webhooks** is a better approach than using HTTP requests:

- Info: https://supabase.com/docs/guides/database/webhooks
- Dashboard: https://supabase.com/dashboard/project/_/integrations/webhooks/webhooks

**Reply by f1refliesss** [reply]
> For new people coming here, I believe **Supabase Database Webhooks** is a better approach than using HTTP requests:
> 
> * Info: https://supabase.com/docs/guides/database/webhooks
> * Dashboard: https://supabase.com/dashboard/project/_/integrations/webhooks/webhooks

But Webhooks are triggered for every action, not just for a specific bucket. Or am I missing something?

I'd like the Webhook to run only when a file is uploaded to a specific bucket (exactly what this issue is about). It seems like that is not possible by using Webhooks?

**Reply by GaryAustin1** [reply]
I would use a trigger and pg_net so you can limit and format the data.  BUT you need to sort the inserts done by upload.  There are two per upload.  The first tests RLS and is rolled back,  then the 2nd actually adds the metadata.
